WIP: keep track of samples per query, set a max # of samples

cstyan opened the pull request:
For #4414 

Mostly looking for feedback on whether or not I'm on the right track here, naive solution for now. 
@brian-brazil @tomwilkie 

Signed-off-by: Callum Stytan <callumstyan@gmail.com>

cstyan commented on promql/engine.go:
```diff
@@ -743,9 +747,14 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
-			sample.Point.T = ts
-			ss.Points = append(ss.Points, sample.Point)
-			seriess[h] = ss
+			if curSamples < maxSamples {
```
it seems in some cases that there's the potential for _no_ samples to be added in the above block of for loops, and so they're then added later down here, so we have to track curSamples here as well

cstyan replied:
Okay, so the output is the correct place to do it then? Or should we limit the input as well? The goal is to reduce the chance of OOM by limit the # of samples returned by a query, in this case are we not doubling up on memory allocation with the `vectors` for the input to the functions `f`, and then building the output?

brian-brazil replied:
You could view it a few ways. It probably doesn't hurt to limit both.

cstyan replied:
So in some cases, the input (or rather, output of `f`) is the only input that we consider when building the output vector. We don't want every function passed to `rangeEval` to have to deal with limiting the # of samples right?

brian-brazil replied:
Yes, only matrix functions and raw lookups should need special handling beyond that

cstyan commented on promql/engine.go:
```diff
@@ -382,7 +384,7 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *EvalStmt) (
 			logger:         ng.logger,
 		}
-		val, err := evaluator.Eval(s.Expr)
+		val, err := evaluator.Eval(s.Expr, ng.maxSamplesPerQuery)
```
Agreed, it's already on my "todo list". I was mostly just exploring how the code was currently working to start.

cstyan commented on promql/engine.go:
```diff
@@ -1116,14 +1149,20 @@ func (ev *evaluator) matrixIterSlice(it *storage.BufferedSeriesIterator, mint, m
 		// Values in the buffer are guaranteed to be smaller than maxt.
 		if t >= mint {
-			out = append(out, Point{T: t, V: v})
+			if curSamples < maxSamples {
```
Failing the query entirely, and returning nothing?

brian-brazil replied:
Yes, same as a timeout.

cstyan replied:
interesting

tomwilkie commented on promql/engine.go:
```diff
@@ -164,14 +164,15 @@ func contextDone(ctx context.Context, env string) error {
 
 // NewEngine returns a new engine.
-func NewEngine(logger log.Logger, reg prometheus.Registerer, maxConcurrent int, timeout time.Duration) *Engine {
+func NewEngine(logger log.Logger, reg prometheus.Registerer, maxConcurrent int, maxSamples int, timeout time.Duration) *Engine {
```
Yes please!

cstyan commented on promql/engine.go:
```diff
@@ -574,11 +578,11 @@ type evaluator struct {
 
 	startTimestamp int64 // Start time in milliseconds.
+	endTimestamp   int64 // End time in milliseconds.
+	interval       int64 // Interval in milliseconds.
```
yeah, wish the diff output was smarter here :+1:

brian-brazil commented on cmd/prometheus/main.go:
```diff
@@ -186,6 +187,8 @@ func main() {
 
 	a.Flag("query.max-concurrency", "Maximum number of queries executed concurrently.").
 		Default("20").IntVar(&cfg.queryConcurrency)
+	a.Flag("query.max-samples", "Maximum number of samples a single query can return. Queries will fail if they would return more than this number.").
```
This message isn't right, it's about samples in memory rather than returned.

Could we express it in bytes? It's a lower bound due to things like labelsets and capacity vs len, but we could improve it over time.

brian-brazil commented on cmd/prometheus/main.go:
```diff
@@ -186,6 +187,8 @@ func main() {
 	a.Flag("query.max-concurrency", "Maximum number of queries executed concurrently.").
 		Default("20").IntVar(&cfg.queryConcurrency)
+	a.Flag("query.max-samples", "Maximum number of samples a single query can return. Queries will fail if they would return more than this number.").
+		Default("1000000").IntVar(&cfg.queryMaxSamples)
```
This should be at least 20M

brian-brazil commented on promql/engine.go:
```diff
@@ -67,13 +68,22 @@ type (
 	ErrQueryTimeout string
 	// ErrQueryCanceled is returned if a query was canceled during processing.
 	ErrQueryCanceled string
+	// ErrTooManySamples is returned if a query would return more than the maximum allowed samples.
```
Samples in memory, not returned

brian-brazil commented on promql/engine.go:
```diff
@@ -668,16 +689,24 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 	matrixes := make([]Matrix, len(exprs))
 	origMatrixes := make([]Matrix, len(exprs))
+	// Limit both input and output samples, in some cases we allocate input vectors.
+	// We'll want to know that the input is too large and fail early rather than also
```
Hmm, shouldn't we already have failed while preparing the inputs?

brian-brazil commented on promql/engine.go:
```diff
@@ -705,11 +734,12 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						inputSamples++
```
That's not right, the input samples are the entire of the input matrix - not jut the bits we've used so far.

cstyan replied:
I'm not sure what you mean, because we already have `matrixes` in memory? We're only appending the samples we end up using to `vectors`, which is then used as `args` to `f`,

brian-brazil replied:
Yes, they're already in memory.

cstyan replied:
So then if the length of all the `Matrix` in `matrixes` is `> maxSamples` should we not just fail right away? Or only if `inputSamples + len(matrixes[i]) > maxSamples`?

brian-brazil replied:
We should have already failed when one of them was being calculated.

cstyan replied:
Right, so if `inputSamples + len matrixes[i] > maxSamples` we should fail, that should happen on line 707

brian-brazil replied:
Yeah around, there. It'd be inside the eval.

cstyan replied:
Well we're already tracking if each call to eval would load more than `maxSamples`, but not if the result of all the calls in rangeEval to eval would.

cstyan replied:
So it seems if we track the # of samples that way, based on the outputs of the eval's, we don't need to track samples where I am here at all. That, in combination with not tracking samples when adding to the output series (https://github.com/prometheus/prometheus/pull/4513/commits/29e1469c81a1476a51c7b5f7ede128bb61ac7005), are we missing any cases? The engine tests I added would all still pass.

brian-brazil commented on promql/engine.go:
```diff
@@ -818,14 +861,19 @@ func (ev *evaluator) eval(expr Expr) Value {
 
 		sel := e.Args[matrixArgIndex].(*MatrixSelector)
-		mat := make(Matrix, 0, len(sel.series)) // Output matrix.
+		mat := make(Matrix, 0, ev.maxSamples) // Output matrix.
```
This would make the matrix way too large, the existing capacity was fine.

brian-brazil commented on promql/engine.go:
```diff
@@ -1112,14 +1190,20 @@ func (ev *evaluator) matrixIterSlice(it *storage.BufferedSeriesIterator, mint, m
 		// Values in the buffer are guaranteed to be smaller than maxt.
 		if t >= mint {
-			out = append(out, Point{T: t, V: v})
+			if curSamples < ev.maxSamples {
```
This should be erroring

cstyan replied:
fixed

brian-brazil commented on promql/engine.go:
```diff
@@ -668,16 +689,21 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
+// step.  The return value is the combination into time series of all the
 // function call results.
-func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ...Expr) Matrix {
+func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ...Expr) (Matrix, error) {
```
You shouldn't need to change this signature, use ev.error

cstyan replied:
we want to panic and recover all error cases? that seems strange to me

brian-brazil replied:
Yes, that's how error handling is done in PromQL.

cstyan commented on promql/engine.go:
```diff
@@ -818,14 +853,19 @@ func (ev *evaluator) eval(expr Expr) Value {
 
 		sel := e.Args[matrixArgIndex].(*MatrixSelector)
 		mat := make(Matrix, 0, len(sel.series)) // Output matrix.
+		curSamples := 0
```
are you referring to this? https://github.com/prometheus/prometheus/blob/master/promql/engine.go#L819-L825

brian-brazil replied:
Yes

brian-brazil commented on promql/engine.go:
```diff
@@ -720,14 +747,19 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 		result := f(args, enh)
 		enh.out = result[:0] // Reuse result vector.
+
+		if len(result) > ev.maxSamples {
```
Yes, both need to be considered.

brian-brazil replied:
Yes, both are present. If there's binary operators, there may also be input vectors from the LHS of binary operators up the call stack.

cstyan commented:
sorry for the delay, I think I'm covering all the cases now but I'll have another look over my PR plus the other PR, I also want to simplify the sample tracking by keeping track of it in the evaluator struct rather than via an int in each function

cstyan commented:
@brian-brazil when you have time please see the latest commit. I've simplified the tracking of the # samples. The # of samples in memory vs the # of samples being returned from an evaluation was tripping me up a bit still, but I think I've got things right this time. I've left comments where I was previously tracking samples but no longer think it's necessary, or where we're not tracking samples in the first place because _I think_ it's not necessary.

I think the metrics around query sizes in the other PR were interesting so I'll add some here. 

If everything's good I'll rebase down to one commit.

brian-brazil commented on promql/engine.go:
```diff
@@ -67,13 +68,22 @@ type (
+	return fmt.Sprintf("query was canceled in %s", string(e))
+}
+func (e ErrTooManySamples) Error() string {
+	return fmt.Sprintf("query would return too many samples in %s", string(e))
```
it's more the processing than the end result

brian-brazil commented on promql/engine.go:
```diff
@@ -677,7 +700,8 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 		// Functions will take string arguments from the expressions, not the values.
 		if e != nil && e.Type() != ValueTypeString {
-			matrixes[i] = ev.eval(e).(Matrix)
+			ret := ev.eval(e)
```
You don't need to change this

cstyan replied:
:+1: sorry, leftover from when I modified the function signatures to return errors

brian-brazil commented on promql/engine.go:
```diff
@@ -705,11 +729,12 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						ev.currentSamples++
```
You need to undo this before the next ts iteration

cstyan replied:
Okay, so in that case I need a query for one of the tests that would load more than a single sample into memory per timestamp. The tests I have at the moment load at most one sample per ts, and I don't think setting the `ev.maxSamples` value to 0 is really a valid test.

brian-brazil replied:
It's more testing across multiple timestamps

cstyan replied:
How so? At the moment none of the tests I've added load more than a single sample into memory per ts.

brian-brazil replied:
The issue is that you're still counting all the inputs from the previous timestamps when doing the current timestamp, but they're no longer in memory.

cstyan replied:
Right, I've fixed that via lines 727/783. What I'm saying is that currently none of the tests have more than a single sample per timestamp. I want to find a query to add for a test that would load say 3 samples per timestamp into memory, and ensure the test fails if we set max samples to 2.

brian-brazil commented on promql/engine.go:
```diff
@@ -705,11 +729,12 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 			for si, series := range matrixes[i] {
 				for _, point := range series.Points {
-					if point.T == ts {
+					if point.T == ts && ev.currentSamples < ev.maxSamples {
```
This should error when it's hit

brian-brazil commented on promql/engine.go:
```diff
@@ -720,6 +745,14 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
+		// in the above for loop that built the input vectors.
+		ev.currentSamples += len(result)
+		if ev.currentSamples > ev.maxSamples {
+			ev.error(error(ErrTooManySamples(env)))
```
Why do you need the `error()` here?

cstyan replied:
do we not have more samples in memory via the return value of `f`?

brian-brazil replied:
I mean specifically, why are you wrapping `ErrTooManySamples` in `error`?

cstyan replied:
Hmm, at some point I had a compile error with this but you're right they're not needed.

brian-brazil commented on promql/engine.go:
```diff
@@ -777,6 +780,7 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 			seriess[h] = ss
 
 		}
+		ev.currentSamples = tempNumSamples
```
This can return early up a bit, so this needs reworking.

The input vectors to the function also need accounting for when they're no longer in use.

cstyan replied:
> The input vectors to the function also need accounting for when they're no longer in use.

Not sure what you mean here. We should reset the number of samples immediately after the call to `f` rather than later on?

brian-brazil replied:
With recursive calls to eval, once the inner evals are done we don't need to account for their memory.

cstyan commented on promql/engine.go:
```diff
@@ -697,10 +697,15 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 		if e != nil && e.Type() != ValueTypeString {
+			tempNumSamples = ev.currentSamples
 			matrixes[i] = ev.eval(e).(Matrix)
+			// Once inner calls to eval are done we don't need to account for memory for
```
Right, the output is saved in `matrixes[i]` and then we're tracking the number of samples in `matrixes` below in the loop for the timestamp. We should be adding `len(matrixes)` to `ev.currentSamples` after resetting to `tempNumSamples` I think.

cstyan replied:
I think this is right.

brian-brazil commented on cmd/prometheus/main.go:
```diff
@@ -186,6 +187,8 @@ func main() {
 	a.Flag("query.max-concurrency", "Maximum number of queries executed concurrently.").
 		Default("20").IntVar(&cfg.queryConcurrency)
+	a.Flag("query.max-samples", "Maximum number of samples a single query can load into memory. Note that queries will fail if they would load more samples than this into memory, so this also limits the number of samples a query can return.").
+		Default("1000000").IntVar(&cfg.queryMaxSamples)
```
I'd suggest 50M, that's nominally 800MB

brian-brazil commented on promql/engine.go:
```diff
@@ -889,12 +944,14 @@ func (ev *evaluator) eval(expr Expr) Value {
 
 	case *UnaryExpr:
-		mat := ev.eval(e.Expr).(Matrix)
+		ret = ev.eval(e.Expr)
```
This can be undone

brian-brazil commented on promql/engine.go:
```diff
@@ -878,8 +927,14 @@ func (ev *evaluator) eval(expr Expr) Value {
-			if len(ss.Points) > 0 {
-				mat = append(mat, ss)
+			// (callum) I think we still need to track number of samples here
+			if len(ss.Points) > 0 && ev.currentSamples < ev.maxSamples {
```
This seems more complicated than it needs to be

brian-brazil commented on promql/engine.go:
```diff
@@ -818,8 +865,10 @@ func (ev *evaluator) eval(expr Expr) Value {
 			if i != matrixArgIndex {
-				otherArgs[i] = ev.eval(e).(Matrix)
+				ret = ev.eval(e)
+				otherArgs[i] = ret.(Matrix)
```
This can be undone

brian-brazil commented on promql/engine.go:
```diff
@@ -668,16 +691,22 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
+			// Once inner calls to eval are done we don't need to account for memory for
+			// samples within those calls. However the result of the call to eval is in mem.
+			ev.currentSamples = tempNumSamples
+			ev.currentSamples += len(matrixes[i])
```
This is the number of series, not the number of samples. I think you're doing the saving of tempNumSamples in the wrong place, and you don't need any code changes to this part as ev.currentSamples will already have the right value.

brian-brazil commented on promql/engine.go:
```diff
@@ -700,16 +729,21 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 	enh := &EvalNodeHelper{out: make(Vector, 0, biggestLen)}
 	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
 	for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Reset num. of samples in memory after each timestamp.
```
But the number of samples will have gone up, due to the output.

brian-brazil commented on promql/engine.go:
```diff
@@ -668,16 +691,22 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 	numSteps := int((ev.endTimestamp-ev.startTimestamp)/ev.interval) + 1
 	matrixes := make([]Matrix, len(exprs))
 	origMatrixes := make([]Matrix, len(exprs))
+	tempNumSamples := 0
```
Shouldn't this be ev.currentSamples? Ultimately when this function returns ev.currentSamples should be the previous value, plus how ever many samples we're returning.

cstyan replied:
wouldn't that just be whatever the value of `ev.currentSamples` is when we return?

cstyan replied:
In every case? After the call to eval, we've potentially added to `ev.currentSamples` in  eval via`*Call`, `*VectorSelector` cases, or via `vectorSelector()`/`matrixSelector()` in which cases we wouldn't have reset `ev.currentSamples` afterwards, or we've called `rangeEval()`.

As I said earlier, I'm sill a bit lost understanding how to properly handle all of the cases. Any pointers would be helpful.

brian-brazil replied:
Yes, in every case. 

> After the call to eval, we've potentially added to ev.currentSamples in ...

Yes, and that needs to be undone before we return as those are no longer in memory.

So if we start a rangeEval with say 10 samples used up by higher in the call stack, then any functions we call we will add to that - so say we're now at 30 samples. If the output of the rangeEval is 5 samples then currentSamples should be 15 by time the rangeEval returns.

cstyan replied:
so then here before a call to `ev.eval` we want to be sure `ev.currentSamples` is saved to a temp var, and after the call `ev.currentSamples` should be set to that temp var plus the length of all the series in `matrixes[i]`?

brian-brazil replied:
Yes

cstyan commented:
Sorry @brian-brazil but I'm still a bit lost trying to understand exactly how the `ev.currentSamples` reset should work.

cstyan commented on promql/engine.go:
```diff
@@ -700,16 +726,21 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 						// Move input vectors forward so we don't have to re-scan the same
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						ev.currentSamples++
```
@brian-brazil so now I assume that since we're adding the # of samples in the matrices returned by the calls to eval, we don't need to add to `ev.currentSamples` here, but we should keep the later block where we add the length of the `result` from `f`. Is that correct?

brian-brazil replied:
No, we still need these as it is a (temporary) 2nd copy of some of the data in the input vector.

cstyan commented on promql/engine.go:
```diff
@@ -720,16 +751,25 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
+
+		// In some cases we only have the result of 'f' without having tracked any samples
+		// in the above for loop that built the input vectors.
+		ev.currentSamples += len(result)
```
keep this?

brian-brazil replied:
Yes, this will be part of the output

What you ultimately want is the original tempNumSamples plus all of these in ev.currentSamples when you return

brian-brazil commented on promql/engine.go:
```diff
@@ -668,16 +691,19 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 	for i, e := range exprs {
 		// Functions will take string arguments from the expressions, not the values.
 		if e != nil && e.Type() != ValueTypeString {
+			tempNumSamples = ev.currentSamples
```
This wipes out the tempNumSamples if there's more than one expr, you don't need this line

cstyan replied:
I don't understand, each call to eval is adding samples to `ev.currentSamples`, but after the call we want the result of `ev.currentSamples` to be the value prior to the `eval` call plus the amount of samples returned.

cstyan replied:
as we discussed here: https://github.com/prometheus/prometheus/pull/4513#discussion_r218979274

brian-brazil replied:
After the call, the value of ev.currentSamples should already be the eval call plus the amount of samples returned - as that's the contract.

brian-brazil commented on promql/engine.go:
```diff
@@ -668,16 +691,19 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 		if e != nil && e.Type() != ValueTypeString {
+			tempNumSamples = ev.currentSamples
 			matrixes[i] = ev.eval(e).(Matrix)
+			ev.currentSamples = tempNumSamples + matrixes[i].TotalSamples()
```
ev.currentSamples will already have the right value due to the eval, no need to change it

brian-brazil commented on promql/engine.go:
```diff
@@ -700,16 +726,21 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
 	for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Reset num. of samples in memory after each timestamp.
+		tempNumSamples = ev.currentSamples
```
You're losing the original value of tempNumSamples here

brian-brazil replied:
You're still losing this here, you need a different variable to store this in.

cstyan replied:
I see what you're saying. The number of samples we have going into the function is the value we want to ues (plus the amount of new samples) when returning, but we want to reset to the previous value of `ev.currentSamples` after each timestamp iteration. Latest commit should fix this.

brian-brazil commented:
No, those changes aren't needed.

The goal should be that at the end of the evalRange call (or any other eval call) that ev.currentSample is the number of samples that function is about to return plus the samples held by all calling functions up the call stack.

cstyan commented:
Okay, I believe the latest commit is correct then.

cstyan commented:
rebased off master and squashed everything but the lastest commit into one commit

brian-brazil commented on promql/engine.go:
```diff
@@ -705,16 +734,21 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
 	for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Reset num. of samples in memory after each timestamp.
+		tempNumSamples = ev.currentSamples
```
This can go above the for, and then you can reset it here - which is more natural than doing it at the end of the loop

brian-brazil commented on promql/engine.go:
```diff
@@ -728,16 +762,25 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
+
+		// In some cases we only have the result of 'f' without having tracked any samples
+		// in the above for loop that built the input vectors.
+		ev.currentSamples += len(result)
```
You want to add this to tempNumSamples as the output data from previous cycles is in memory.

cstyan replied:
Hmmm, sorry I don't understand why. 

In addition to, or instead of adding it to `ev.currentSamples`? It seems to me that it should be in addition to since we check `ev.currentSamples` when determining if we have too many samples in memory.

brian-brazil replied:
Yes. After every iteration we're now holding the results of previous iterations in memory, which needs to be part of tempNumSamples.

brian-brazil commented on promql/engine.go:
```diff
@@ -728,16 +763,29 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 		}
 		enh.out = result[:0] // Reuse result vector.
+
+		// In some cases we only have the result of 'f' without having tracked any samples
```
I don't think this comment helps, one way or the other we should always be adding this.

brian-brazil commented on promql/engine.go:
```diff
@@ -1127,14 +1200,22 @@ func (ev *evaluator) matrixIterSlice(it *storage.BufferedSeriesIterator, mint, m
 		t, v := it.Values()
 		if t == maxt && !value.IsStaleNaN(v) {
-			out = append(out, Point{T: t, V: v})
+			if ev.currentSamples < ev.maxSamples {
```
This should error

brian-brazil commented on promql/engine.go:
```diff
@@ -331,9 +352,8 @@ func (ng *Engine) exec(ctx context.Context, q *query) (Value, error) {
-
-	evalSpanTimer, ctx := q.stats.GetSpanTimer(ctx, stats.EvalTotalTime)
-	defer evalSpanTimer.Finish()
+	evalTimer := q.stats.GetTimer(stats.EvalTotalTime).Start()
```
Why does the function change here?

cstyan replied:
I think I picked the wrong change when fixing a merge conflict here during rebase, fixing.

brian-brazil commented on promql/engine.go:
```diff
@@ -704,17 +734,23 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 	seriess := make(map[uint64]Series, biggestLen) // Output series by series hash.
+	tempNumSamples = ev.currentSamples
 	for ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval {
+		// Reset num. of samples in memory after each timestamp.
```
number

brian-brazil commented on promql/engine.go:
```diff
@@ -673,15 +697,21 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 	origMatrixes := make([]Matrix, len(exprs))
+
+	originalNumSamples := ev.currentSamples
+
```
There's a bit much vertical whitespace here

brian-brazil commented on promql/engine.go:
```diff
@@ -673,15 +697,21 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
+
+	originalNumSamples := ev.currentSamples
+
+	var tempNumSamples int
```
Why not create this when you assign to it?

brian-brazil commented on promql/engine.go:
```diff
@@ -704,17 +734,23 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 						// past points at the next step.
 						matrixes[i][si].Points = series.Points[1:]
+						ev.currentSamples++
+					} else if ev.currentSamples >= ev.maxSamples {
```
The else here and duplicated inverse condition is a bit odd, it seems like it might be accidentally broken bu a future change. I'd move this inside the above if as a sub if.

brian-brazil commented on promql/engine.go:
```diff
@@ -728,16 +764,27 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
+			ev.currentSamples = originalNumSamples + mat.TotalSamples()
 			return mat
 		}
-		// Add samples in output vector to output series.
```
This should stay

brian-brazil commented on promql/engine.go:
```diff
@@ -828,6 +878,7 @@ func (ev *evaluator) eval(expr Expr) Value {
 			if i != matrixArgIndex {
 				otherArgs[i] = ev.eval(e).(Matrix)
 				otherInArgs[i] = Vector{Sample{}}
+				// (callum) this is just building args for later procesing so I don't think we need to track samples here
```
Can you either clean these up or make them proper comments?

cstyan commented:
removed most of the comments you pointed out as they were only necessary while I was trying to track where samples were loaded in and out of memory, updated the rest and simplified the if block

brian-brazil commented:
Thanks!

maozi07 commented:
I think it's better to check the memory prometheus used and have a limit, not the amount of samples. In my situation that's caused some problems for a max sample.Thank you all for your work

ernanirst commented:
@brian-brazil I read that each sample takes 16 bytes, so if I'm using a VM that has 8GB of RAM available, how would you advice on calculating `query.max-samples`? I'm supposing I can consider 1 GB for OS, 1 GB for Prometheus itself and the remaining 6 GB to the samples? And therefore 6GB / 16 bytes = 3,75 * 10^8 samples?

 Does this math makes sense?

ringerc commented:
This limits sample counts, but doesn't do any accounting for the labels for each series, right? 

From what I'm seeing in testing, loading a lot of "wide" series due to queries that consume a lot of info-metrics and/or query some high-churn info-metric series can still easily OOM the server even with quite conservative sample limits.
